File: packages/domain/gateway.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from uuid import UUID

from packages.domain.models import ModelProfile


class Priority(str, Enum):
    BALANCED = "balanced"
    LOW_COST = "low_cost"
    HIGH_QUALITY = "high_quality"
    LOW_LATENCY = "low_latency"


class ModelTier(str, Enum):
    TIER1_VERIFIED = "tier1_verified"
    TIER2_PROVISIONAL = "tier2_provisional"


class HealthState(str, Enum):
    HEALTHY = "healthy"
    DEGRADED = "degraded"
    BROKEN = "broken"


class Intent(str, Enum):
    GENERAL = "general"
    ANALYSIS = "analysis"
    CODE = "code"
    CREATIVE = "creative"


def intent_from_evaluation_string(value: str) -> Intent:
    """Map prompt-evaluator intent labels to domain Intent."""
    key = value.strip().lower()
    mapping = {
        "code": Intent.CODE,
        "analysis": Intent.ANALYSIS,
        "creative": Intent.CREATIVE,
        "general": Intent.GENERAL,
    }
    return mapping.get(key, Intent.GENERAL)


@dataclass(frozen=True)
class GatewayTask:
    prompt: str
    priority: Priority
    temperature: float | None
    max_tokens: int | None
    require_json: bool
    max_cost_usd: float | None = None
    discovery_mode: bool = False
    simulate_failures: list[str] = field(default_factory=list)
    use_cases: list[str] = field(default_factory=list)
    preferred_providers: list[str] = field(default_factory=list)
    response_depth: str = "balanced"


@dataclass(frozen=True)
class RoutedRequest:
    prompt: str
    temperature: float
    max_tokens: int | None
    require_json: bool
    simulate_failures: set[str] = field(default_factory=set)


@dataclass(frozen=True)
class ProviderResponse:
    content: str
    provider: str
    model_id: str
    input_tokens: int | None = None
    output_tokens: int | None = None
    latency_ms: int | None = None
    cost: float = 0.0


@dataclass(frozen=True)
class InvocationAttempt:
    provider: str
    model_id: str
    status: str
    detail: str
    latency_ms: int | None = None


@dataclass(frozen=True)
class RankingHighlight:
    """One model slot in the user-facing ranking summary (not debug scoring)."""

    model_id: str
    display_name: str
    provider: str
    reason_key: str
    same_as_best_overall: bool = False


@dataclass(frozen=True)
class RankingSummary:
    best_overall: RankingHighlight
    free_alternative: RankingHighlight | None
    best_quality: RankingHighlight
    best_cost: RankingHighlight
    best_speed: RankingHighlight


@dataclass(frozen=True)
class ScoredCandidate:
    model: ModelProfile
    priority_weight: int
    db_model_id: int
    rank: int
    quality_score: float
    latency_score: float
    cost_score: float
    final_score: float
    model_score_adjustment: float
    explanation: str
    pros: tuple[str, ...]
    cons: tuple[str, ...]
    tier: ModelTier = ModelTier.TIER2_PROVISIONAL
    health_status: HealthState = HealthState.HEALTHY
    snapshot_latency_p50: float | None = None
    user_rating: float | None = None
    user_rating_count: int = 0


@dataclass(frozen=True)
class RoutingDecision:
    intent: Intent
    reason: str
    applied_temperature: float
    candidates: list[ModelProfile]
    scored_candidates: tuple[ScoredCandidate, ...]
    preferred_providers: list[str] = field(default_factory=list)
    preferred_providers_applied: bool = False
    preferred_providers_fallback_used: bool = False


@dataclass(frozen=True)
class FallbackExecutionOutcome:
    response: ProviderResponse
    attempts: list[InvocationAttempt]


@dataclass(frozen=True)
class GatewayExecutionResult:
    request_id: UUID
    response: ProviderResponse
    decision: RoutingDecision
    attempts: list[InvocationAttempt]
    fallback_used: bool
    ranking_summary: RankingSummary


class NoModelsAvailableError(RuntimeError):
    """Raised when there are no database routing candidates for the request."""

File: apps/web/src/shared/api/gateway.ts
import { getSessionId } from "./session"

const API_BASE = import.meta.env.PUBLIC_API_BASE_URL as string

type PriorityMapping = Record<string, string>

const PRIORITY_MAP: PriorityMapping = {
  quality: "high_quality",
  speed: "low_latency",
  cost: "low_cost",
}

export type GatewayAdvancedOptions = {
  useCases?: string[]
  preferredProviders?: string[]
  responseDepth?: "short" | "balanced" | "detailed"
}

export type RankingHighlightBackend = {
  model_id: string
  display_name: string
  provider: string
  reason_key: string
  same_as_best_overall: boolean
}

export type RankingSummaryBackend = {
  best_overall: RankingHighlightBackend
  free_alternative: RankingHighlightBackend | null
  best_quality: RankingHighlightBackend
  best_cost: RankingHighlightBackend
  best_speed: RankingHighlightBackend
}

export type GatewayBackendResponse = {
  request_id: string
  content: string
  provider: string
  model_id: string
  recommended_model_id: string
  response_latency_ms: number | null
  intent: string
  priority: string
  routing_reason: string
  explanation: string
  ranking_summary: RankingSummaryBackend
  ranking: Array<{
    model_id: string
    rank: number
    quality_score: number
    latency_score: number
    cost_score: number
    final_score: number
    model_score_adjustment: number
    explanation: string
    pros: string[] | null
    cons: string[] | null
    context_window: number | null
    max_output_tokens: number | null
    supports_json: boolean
    supports_tools: boolean
    model_categories?: string[]
    technical_capabilities?: string[]
    verification_scopes?: string[]
    capabilities: string[]
    is_free: boolean
    tier: string
    evaluation_status?: string
    supports_vision?: boolean
    input_modalities?: string[]
    output_modalities?: string[]
    model_type_labels?: string[]
    is_verified?: boolean
    public_status_key?: string | null
    user_rating?: number | null
    user_rating_count?: number
    cost_per_million_input?: number | null
    cost_per_million_output?: number | null
  }>
  fallback_used: boolean
  candidate_models: string[]
  attempts: Array<{
    provider: string
    model_id: string
    status: string
    detail: string
    latency_ms: number | null
  }>
  preferred_providers: string[]
  preferred_providers_applied: boolean
  preferred_providers_fallback_used: boolean
}

export async function callGateway(
  prompt: string,
  priority: string,
  options?: GatewayAdvancedOptions,
  signal?: AbortSignal,
): Promise<GatewayBackendResponse> {
  const res = await fetch(`${API_BASE}/v1/chat/completions/advanced`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-Session-Id": getSessionId(),
    },
    body: JSON.stringify({
      prompt,
      priority: PRIORITY_MAP[priority] ?? "balanced",
      max_tokens: null,
      use_cases: options?.useCases ?? [],
      preferred_providers: options?.preferredProviders ?? [],
      response_depth: options?.responseDepth ?? "balanced",
    }),
    signal,
  })
  if (!res.ok) throw new Error(`Gateway ${res.status}`)
  return res.json() as Promise<GatewayBackendResponse>
}
